packages/core/contracts/financial-templates/perpetual-poolParty/PerpetualPositionManagerPoolPartyLib.sol

```diff
@@ -0,0 +1,596 @@
+
+        emit Repay(msg.sender, numTokens.rawValue, newTokenCount.rawValue);
+
+        // Transfer the tokens back from the sponsor and burn them.
+        positionManagerData.tokenCurrency.safeTransferFrom(msg.sender, address(this), numTokens.rawValue);
+        positionManagerData.tokenCurrency.burn(numTokens.rawValue);
+    }
+
+    function settleEmergencyShutdown(
+        PerpetualPositionManagerPoolParty.PositionData storage positionData,
+        PerpetualPositionManagerPoolParty.GlobalPositionData storage globalPositionData,
+        PerpetualPositionManagerPoolParty.PositionManagerData storage positionManagerData,
+        FeePayerPoolParty.FeePayerData storage feePayerData
+    ) external returns (FixedPoint.Unsigned memory amountWithdrawn) {
+        if (positionManagerData.emergencyShutdownPrice.isEqual(FixedPoint.fromUnscaledUint(0))) {
+            FixedPoint.Unsigned memory oraclePrice = positionManagerData._getOracleEmergencyShutdownPrice(feePayerData);
+            positionManagerData.emergencyShutdownPrice = oraclePrice._decimalsScalingFactor(feePayerData);
+        }
+
+        // Get caller's tokens balance and calculate amount of underlying entitled to them.
+        FixedPoint.Unsigned memory tokensToRedeem =
+            FixedPoint.Unsigned(positionManagerData.tokenCurrency.balanceOf(msg.sender));
+
+        FixedPoint.Unsigned memory totalRedeemableCollateral =
+            tokensToRedeem.mul(positionManagerData.emergencyShutdownPrice);
+
+        // If the caller is a sponsor with outstanding collateral they are also entitled to their excess collateral after their debt.
+
+        if (
+            positionData.rawCollateral.getFeeAdjustedCollateral(feePayerData.cumulativeFeeMultiplier).isGreaterThan(0)
+        ) {
+            // Calculate the underlying entitled to a token sponsor. This is collateral - debt in underlying with
+            // the funding rate applied to the outstanding token debt.
+
+            FixedPoint.Unsigned memory tokenDebtValueInCollateral =
+                positionData.tokensOutstanding.mul(positionManagerData.emergencyShutdownPrice);
+            FixedPoint.Unsigned memory positionCollateral =
+                positionData.rawCollateral.getFeeAdjustedCollateral(feePayerData.cumulativeFeeMultiplier);
+
+            // If the debt is greater than the remaining collateral, they cannot redeem anything.
+            FixedPoint.Unsigned memory positionRedeemableCollateral =
+                tokenDebtValueInCollateral.isLessThan(positionCollateral)
+                    ? positionCollateral.sub(tokenDebtValueInCollateral)
+                    : FixedPoint.Unsigned(0);
+
+            // Add the number of redeemable tokens for the sponsor to their total redeemable collateral.
+            totalRedeemableCollateral = totalRedeemableCollateral.add(positionRedeemableCollateral);
+
+            PerpetualPositionManagerPoolParty(address(this)).deleteSponsorPosition(msg.sender);
```

Comment:
Why not just pass in positions and delete it inline rather than doing this callback?

Reply:
yes, the idea of pass mapping in struct is another option. we decide for this in order to make the library more light, since all the mappings are called at the beginning and the only next interaction is the delete